client/src/procedures/primitives.rs

```diff
@@ -55,108 +43,204 @@ pub enum PrimitiveProcedure {
+    // Stronghold procedures that implement the `UseSecret` trait.
+    UseSecret => { PublicKey, Ed25519Sign, Hmac, AeadEncrypt, AeadDecrypt },
+    // Stronghold procedures that directly implement the `Procedure` trait.
+    _ => { RevokeData, GarbageCollect }
+}
+
+/// Write data to the specified [`Location`].
+#[derive(Clone, GuardDebug, Serialize, Deserialize)]
+pub struct WriteVault {
+    pub data: Vec<u8>,
+
+    pub location: Location,
+
+    pub hint: RecordHint,
+}
+
+impl GenerateSecret for WriteVault {
+    type Output = ();
+
+    fn generate(self) -> Result<Products<Self::Output>, FatalProcedureError> {
+        Ok(Products {
+            secret: self.data,
+            output: (),
+        })
+    }
+
+    fn target(&self) -> (&Location, RecordHint) {
+        (&self.location, self.hint)
+    }
+}
+
+/// Revoke the data from the specified [`Location`]. Revoked data is not readable and can be
+/// removed from a vault with the [`GarbageCollect`] Procedure. If the `should_gc` flag is set to `true`,
+/// it with automatically cleanup the revoke. Otherwise, the data is just marked as revoked.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct RevokeData {
+    pub location: Location,
+    pub should_gc: bool,
+}
+
+impl Procedure for RevokeData {
+    type Output = ();
+    fn execute<R: Runner>(self, runner: &mut R) -> Result<Self::Output, ProcedureError> {
+        runner.revoke_data(&self.location)?;
+        if self.should_gc {
+            runner.garbage_collect(SecureClient::resolve_location(self.location).0);
         }
+        Ok(())
     }
 }
 
-// === implement From Traits from inner types to wrapper enums
-
-enum_from_inner!(PrimitiveProcedure::CopyRecord from CopyRecord);
-enum_from_inner!(PrimitiveProcedure::Slip10Generate from Slip10Generate);
-enum_from_inner!(PrimitiveProcedure::Slip10Derive from Slip10Derive);
-enum_from_inner!(PrimitiveProcedure::BIP39Generate from BIP39Generate);
-enum_from_inner!(PrimitiveProcedure::BIP39Recover from BIP39Recover);
-enum_from_inner!(PrimitiveProcedure::GenerateKey from GenerateKey);
-enum_from_inner!(PrimitiveProcedure::PublicKey from PublicKey);
-enum_from_inner!(PrimitiveProcedure::Ed25519Sign from Ed25519Sign);
-enum_from_inner!(PrimitiveProcedure::X25519DiffieHellman from X25519DiffieHellman);
-enum_from_inner!(PrimitiveProcedure::Hash from Hash);
-enum_from_inner!(PrimitiveProcedure::Hmac from Hmac);
-enum_from_inner!(PrimitiveProcedure::Hkdf from Hkdf);
-enum_from_inner!(PrimitiveProcedure::Pbkdf2Hmac from Pbkdf2Hmac);
-enum_from_inner!(PrimitiveProcedure::AeadEncrypt from AeadEncrypt);
-enum_from_inner!(PrimitiveProcedure::AeadDecrypt from AeadDecrypt);
-
-// ==========================
-// Helper Procedure
-// ==========================
+/// Garbage collects any revokes in a Vault based on the given `vault_path`.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GarbageCollect {
```

Comment:
> but this could be part of a complete procedure call to automatically clean up unused locations

@felsweg-iota  what do you mean by  "unused locations"?

Reply:
sorry, wrong wording in this case. I am referring to locations that have been soft-deleted.

Reply:
@felsweg-iota what do you mean by "soft-deleted"? 😅 Could you give an example?

Reply:
"Soft-Deletion" is a term used in databases to indicate a record to be deleted, while effectively it is not. A record (inside a db table) gets a marker field, where the time of deletion is recorded, the record is thus revoked. At a later time, the record gets effectively deleted from the table.

Having said that, Stronghold also has an atomic pair of `revoke` / `garbage collect` operations. `delete_data` (in the present interface) does exactly that. A boolean value indicates either just `revokation` of data, or direct deletion.

Reply:
So if I understand you correctly, you are proposing that after the procedure call we automatically run garbage-collect on the vaults where data has been revoked during the procedure call?

Reply:
yes

Reply:
Okay I think that's a good idea

> Having said that, Stronghold also has an atomic pair of revoke / garbage collect operations. delete_data (in the present interface) does exactly that. A boolean value indicates either just revokation of data, or direct deletion.

The new `RevokeData` procedure has this flag as well; if `should_gc` it garbage collects the vault from which the data was deleted. 

What I am wondering is whether we should just change `RevokeData` to always directly delete the record. 
I don't really understand the benefit in the flow of:
1. revoke-data: creating a `RevokeTransaction` for a record 
2. garbage-collect: iterate through all records from a vault and remove the records with such a `RevokeTransaction`

rather than just removing the record from the `HashMap` of records. 
@tensor-programming could you expand a bit on what the initial motivation was behind this? Do you see a reason not to just remove the record completely when revoking it?